Blog post codestarts preview (#704)
quarkusio/quarkusio.github.io: _posts/2020-08-20-codestarts-preview.adoc

---
layout: post
title: 'Codestarts - Preview the new way to start coding with Quarkus'
date: 2020-08-20
tags: extension codestart quickstart

> synopsis: The idea is simple, you pick your extensions and they are able to provide example code to get you started with it :)

Reviewer: ```suggestion
synopsis: The idea is simple, you pick your extensions and they are able to provide example code to get you started.
```

author: adamevin
---

= Preview the new way to start coding with Quarkus

*Have you ever generated a Quarkus project using the Maven plugin or code.quarkus.io?*

If you did, then you might have wondered why, with the breadth of Quarkus extensions,
we would only let you generate a REST API example application. Of course,
you can pick extensions which are going to be in the dependencies of your app,

> this is already pretty cool because some extensions are already providing new features without any extra code. However,

Reviewer 2: ```suggestion
this is pretty cool because some extensions are providing new features without any extra code. However,
```

some extensions also require configuration and/or code to work, making it pretty limited as it is.

== Introducing our new code generation system

> This is why I am thrilled to introduce the preview of our new code generation system in Quarkus `1.7.0.Final`.

Reviewer: ```suggestion
This is why I am thrilled to introduce the preview of our new code generation system in Quarkus 1.7.0.Final.
```

> This is why I am thrilled to introduce the preview of our new code generation system in Quarkus `1.7.0.Final`.

Reviewer: I usually use the `we` rather than the `I`.

Author: Changed..

> It is called Codestart. The idea is simple, you pick your extensions and they are able to provide example code to get you started with it :)

Reviewer: ```suggestion
It is called Codestart. The idea is simple, you pick your extensions and they are able to provide example code to get you started.
```

> This preview version is just the beginning and includes only 3 example codestarts:

Reviewer 2: ```suggestion
This preview version is just the beginning and includes three example codestarts:
```

> - The legendary RestEasy example in Java, Kotlin and Scala, when you pick the RestEasy extension

Reviewer: ```suggestion
- The legendary RESTEasy example in Java, Kotlin and Scala, when you pick the RESTEasy extension
```

> - The legendary RestEasy example in Java, Kotlin and Scala, when you pick RestEasy extension

Reviewer 2: ```suggestion
- The legendary RestEasy example in Java, Kotlin and Scala, when you pick the RestEasy extension
```

> - The lovely and sweet template engine Qute in Java, when you pick Qute extension

Reviewer 2: ```suggestion
- The lovely and sweet template engine Qute in Java, when you pick the Qute extension
```

- The tough and strong Quarkus Command Mode in Java, when you don’t pick any extension that provides an example

> 

Reviewer 2: legendary, lovely and sweet, tough and strong :) awesome.

Author: :)

== Wanna give it a try?

We added a feature flag in the Quarkus Maven plugin to create your application:
[source,shell,subs=attributes+]
----

> mvn io.quarkus:quarkus-maven-plugin:create -DcodestartsEnabled`

Reviewer: You say it's called `Codestart` singular above but the option is called `codestarts` with an s?

Author: You are right, `-DCodestartEnabled` would have been better.

Reviewer: Well, you call it Codestarts elsewhere in the document so that's not a big issue.

Author: I would say that, "Codestart" is the feature, then each item is a "codestart", that can be plural, the feature is not supposed to be used as plural, but now I guess it's too late for that..

Author: then let's say the feature is `Codestarts`, I make sure the article is consistent and rebase.

Author: Ok, rebased. So `Codestarts` is the feature with the First letter upper cap, codestarts means a group of codestart items, a codestart is just one (without cap).

----

=== A Command mode Quarkus project using Maven as build tool:
[source,shell,subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:create -DcodestartsEnabled \

> -DprojectGroupId=org.acme -DprojectArtifactId=go-go-commando

Reviewer 2: is projectGroupid even needed ? 

always felt like we were being overly cautious on requiring users to list all the arugments when defaults could do.

Author: If you don't it's auto asked stupid :-/

./mvnw package
java -jar target/go-go-commando-1.0-SNAPSHOT-runner.jar
----

===  Same as previous using Gradle as build tool:
[source,shell,subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:create -DcodestartsEnabled \
-DprojectGroupId=org.acme -DprojectArtifactId=go-go-commando-gradle \
-DbuildTool=GRADLE
cd go-go-commando-gradle
# ./gradlew.bat on Window
./gradlew build
java -jar build/go-go-commando-gradle-1.0-SNAPSHOT-runner.jar
----

=== A Quarkus project in Kotlin with example for RESTEasy and Qute:
[source,shell,subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:create -DcodestartsEnabled \
-DprojectGroupId=org.acme -DprojectArtifactId=kotlin-with-rest-is-qute \
-Dextensions=kotlin,resteasy,qute
cd kotlin-with-rest-is-qute
# ./mvnw.cmd on Windows
./mvnw quarkus:dev
----

=== A blank Quarkus project in Scala with SmallRye Reactive Messaging Kafka
[source,shell,subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:create -DcodestartsEnabled \
-DprojectGroupId=org.acme -DprojectArtifactId=scala-kafka \
-Dextensions=scala,quarkus-smallrye-reactive-messaging-kafka \
-DskipExample
cd scala-kafka
# ./mvnw.cmd on Windows
./mvnw quarkus:dev
----

> NOTE: We added instruction to get you started (contextual to the extensions and build tool you've selected) in the new project `README.md` .

Reviewer: ```suggestion
NOTE: We added instructions to get you started (contextual to the extensions and build tool you've selected) in the new project `README.md` .
```

== Start a blank Quarkus application with your favorites tools/extensions

Codestarts also introduce a huge improvement 😅 and don’t laugh, it has been a legitimate demand from the community for a while now,
using codestarts, you can generate a blank Quarkus app by adding: `-DskipExample`.

== You might notice a few missing features with Codestarts

1. The package name of the generated examples is now fixed to `org.acme`,

> we thought it wouldn’t add much value to make it customizable, but let us know if you find it’s a loss.

Reviewer: Seriously? This is an issue IMHO. People obviously want a custom package and the first thing they will have to do is to rename the package.

Author: @gsmet this is all example code meant to be deleted, or used as example. We are not generating the user application, but showing example of what they can use in there app.. So they will have to pick what they want and do their cooking with it..

Author: @gsmet you can still set the group id for maven/gradle. For the package/class, the convention to use in example codestarts to avoid conflicts is:
`org.acme.[example-name].MyClass`

Also think that before you were able to set the class name (not just the package), now you have multiple examples, this feature is not really making sense anymore. IMO if this we didn't have an existing codegen with that feature (and I wouldn't have noted it), you wouldn't even have noticed :)

Author: The `classname` and `path` options were making sense because the current codegen is generating an application to start from. I would say this is required if we were implementing an openapi code generator for example, but this is not what we are trying to achieve here, those are just bits of code to get inspiration.

Reviewer 2: lets open an issue on this if we feel strongly its needed (where we can discuss all the hoops we would need to jump through to support it - its more than just a prefix + folder rename). Shouldn't stop the blog to go out as this is the current situation.

Author: we already have an issue: https://github.com/quarkusio/quarkus/issues/11154

> 2. Since a project will contain multiple examples, we must avoid conflicts, this is why you can't define the REST `path` or your app anymore.

Reviewer 2: ```suggestion
2. Since a project will contain multiple examples, we must avoid conflicts, this is why you can't define the REST `path` in your app anymore.
```
not sure if 'in' or 'of' or you actually meant some other settings besides path ?

== BUGS WANTED

> Codestarts are still in preview mode in Quarkus `1.7.x` for the sake of stabilization but we would like to make the default behavior as soon as possible..

Reviewer: ```suggestion
Codestarts are still in preview mode in Quarkus 1.7.x for the sake of stabilization but we would like to make it the default behavior as soon as possible.
```
